### DISCUSSION TITLE
Unexpected tween behaviour

### DISCUSSION THREAD

**Discussion by danfriedz** [ramokz/phantom-camera#153] [Question / Help Wanted]
https://github.com/ramokz/phantom-camera/assets/61813510/d5233923-aa8c-422c-9f0f-f3a451abe36a

https://github.com/ramokz/phantom-camera/assets/61813510/174e689b-2a04-40cf-a1d4-f84d2db33924

I've included two videos, firstly of me switching between two cameras manually in editor with a 1 sec tween.
The other video I made the tween 5 seconds to show how far away it starts from the player.

From the looks of it, it appears that the camera is tweening from the scenes (0,0,0) point.
How would I make it so it only ever tweens from the previous camera's location?

**Comment by ramokz** [comment]
What you're asking for is the intended behaviour, so what you're seeing is definitely not meant to happen.

There was a similar report (#136) recently about tweening occurring from an unexpected start position.
The initial report was due to an incorrect positional getter in the `PCamHost` script, which should be fixed in the main branch. 

To quickly see if it's the same issue here, does the parent nodes of the `Camera3D`, i.e. the `Ship Container` and `Test_Space_Ship`, have a `Position` that is **not** 0, 0, 0?

**Comment by danfriedz** [comment]
Grabbing the latest version fixed the bug I was having, Thanks! it must have been the same issue.
Although I am having some other tweening issues.

1. tween seems to change depending on the instantiated scenes orientation in the world.
In this video the tween appears to work correctly in one direction. but preforms differently when the ship is facing a different direction or isn't level with the ground. 

https://github.com/ramokz/phantom-camera/assets/61813510/ee5c3aa5-6853-4187-b5f4-eca0a22bd95e

2. When the ship is moving, the tween path seems to be relevant to the world, not the ship. Although I could see this being intended behavior.

I tested this at slow speeds and the camera still leaves the ship behind noticeably.. 

https://github.com/ramokz/phantom-camera/assets/61813510/af4f4860-8176-4bf2-aac5-fcc677e4edb7

any advice (should I open these as separate issues?)

**Reply by ramokz** [reply]
To the first point, that does look like a bug, so opening an issue for that would be great!
I can try and reproduce your setup, but if you have a way to share a simple scene that reflects yours, then that would make it easier to debug. Am also noticing some minor snaps happening at the end of the first tween example, which shouldn't happen either, so am also curious about what's causing that.

For the second point, you're correct that it is technically an intended behaviour of the tweening system. As the `Camera3D` is essentially moving from A → B in global space, where if B is rapidly moving during the tween it will stretch the distance and will make the `Camera3D` exponentially catch up over the course of the given tween's duration.
Would imagine the simplest solve would be to move a single `PCam3D` between the two positions, e.g. by using the Godot [tween method](https://docs.godotengine.org/en/latest/classes/class_tween.html). Though it's not ideal since it effectively eliminates the point of the addon's tweening system.
All that said, I am wondering if there's a way to allow your scene to work correctly out-of-the-box. Think it's definitely worth exploring / testing further, as having fast moving `PCam` position targets does not seem like an uncommon situation.
Would suggest rasing a separate issue just to have a place for that conversation and, potentially, to be addressed in.

**Comment by danfriedz** [comment]
Hi,
Hope you are having a nice new years / christmas etc.
I will open an issue for the first point. Been a little busy lately.

For the second point I took your advice and used the tweened the transforms between some preset pCam transforms. Tweening now occurs in local space and it solved my first issue too.

For anyone looking for a solution in the meantime here is how I am handling tweening currently. 
![image](https://github.com/ramokz/phantom-camera/assets/61813510/5ffaf7e4-6b96-4253-b239-ac43d2ac866e)
Note that i'm just tweening the transform of "Main_Phantom_Camera" to the transform of whatever phantom camera i like. "Main_Phantom_Camera" is always active.
This allows me to bypass my tweening issues while still allowing use of phantoms cameras other functionally (e.g. look and follow modes)

**Comment by NicolaDau** [comment]
Hi @ramokz!

Has the first issue been fixed? I seem to be having the same problem on 0.8.

**Reply by ramokz** [reply]
Could you be more specific? Do you mean the rotation of the tween is not being consistent?

**Reply by NicolaDau** [reply]
Sorry! :) 

Yeah I meant the first issue mentioned by danfriedz, the rotation of the tween not being consistent based on the instantiated scenes' orientation in the world

**Reply by ramokz** [reply]
The rotation of the tween should be automatic, in the sense that it rotates using the shortest route between its initial and final destination.
